### PR TITLE
embeddings benchmarking extended for tei, infinity and rerank

### PR DESCRIPTION
**Pull request by dtrawins**
### 🧪 Checklist

- [ ] Unit tests added.
- [ ] The documentation updated.
- [ ] Change follows security best practices.
``

### CODE REVIEW

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 1]
```diff
@@ -37,19 +37,24 @@
 parser = argparse.ArgumentParser(description='Run benchmark for embeddings endpoints', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
 parser.add_argument('--dataset', required=False, default='Cohere/wikipedia-22-12-simple-embeddings', help='Dataset for load generation from HF or a keyword "synthetic"', dest='dataset')
 parser.add_argument('--synthetic_length', required=False, default=510, type=int, help='Length of the synthetic dataset', dest='length')
-parser.add_argument('--api_url', required=False, default='http://localhost:8000/v3/embeddings', help='API URL for embeddings endpoint', dest='api_url')
+parser.add_argument('--api_url', required=False, help='API URL for embeddings endpoint', dest='api_url')
```
Comment (mzegla) [inline demos/benchmark/embeddings/benchmark_embeddings.py:40]: If there's no default then maybe this parameter should be required?

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 2]
```diff
@@ -37,19 +37,24 @@
...
 args = vars(parser.parse_args())
+
+backend_function = None
+default_api_url = None
+
 docs = Dataset.from_dict({})
 if args["dataset"] == 'synthetic':
     dummy_text = "hi " * args["length"]
-    for i in range(1000):
+    for i in range(100):
```
Comment (mzegla) [inline demos/benchmark/embeddings/benchmark_embeddings.py:54]: Looks like something to parametrize.

File: demos/benchmark/embeddings/benchmark_embeddings.py (hunk 1 above)
Comment (dtrawins) [reply demos/benchmark/embeddings/benchmark_embeddings.py:40]: the default is set depending on the serving backend

File: demos/rerank/README.md [hunk 3]
```diff
@@ -87,11 +89,32 @@ curl http://localhost:8000/v2/rerank  -H "Content-Type: application/json" \
...
+pip3 install cohere
+```
+```bash
+echo '
+import cohere
+client = cohere.Client(base_url="http://localhost:8000/v3", api_key="not_used")
+responses = client.rerank(query="hello",documents=["welcome","farewell"], model="BAAI/bge-reranker-large")
+for response in responses.results:
+    print("index {}, relevance_score {}".format(response.index, response.relevance_score)) > rerank_client.py
+
```
Comment (porlows1) [inline demos/rerank/README.md:91]: missing ' to end echo text

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 4]
```diff
@@ -37,19 +37,26 @@
...
+
 docs = Dataset.from_dict({})
 if args["dataset"] == 'synthetic':
     dummy_text = "hi " * args["length"]
-    for i in range(1000):
+    for i in range(args["limit"]):
         docs = docs.add_item({"text":dummy_text})
 else:
-    docs = load_dataset(args["dataset"],split='train[:1000]')
+    filter = "train[:{}]".format(args["limit"])
```
Comment (pgladkows) [inline demos/benchmark/embeddings/benchmark_embeddings.py:58]:
```suggestion
    filter = f"train[:{args["limit"]}]"
```

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 5]
```diff
@@ -122,6 +129,94 @@ async def async_request_embeddings(
...
+                            continue
+                        #chunk_bytes = chunk_bytes.decode("utf-8")
+                        # data = json.loads(chunk_bytes)
+                        timestamp = time.perf_counter()
+                        output.success = True
+                        output.latency =  timestamp - st
+                else:
+                    output.error = response.reason or ""
+                    output.success = False
+                    print("ERROR",response.reason)
```
Comment (pgladkows) [inline demos/benchmark/embeddings/benchmark_embeddings.py:166]:
```suggestion
                    print("ERROR", response.reason)
```

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 6]
```diff
@@ -122,6 +129,94 @@ async def async_request_embeddings(
...
+        st = time.perf_counter()
+        try:
+            async with session.post(url=api_url, json=payload,
+                                    headers=headers) as response:
+                if response.status == 200:
+                    async for chunk_bytes in response.content:
+                        if not chunk_bytes:
+                            continue
+                        #chunk_bytes = chunk_bytes.decode("utf-8")
+                        # data = json.loads(chunk_bytes)
```
Comment (pgladkows) [inline demos/benchmark/embeddings/benchmark_embeddings.py:202]:
```suggestion
```

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 7]
```diff
@@ -165,7 +260,23 @@ async def limited_request_func(request_func_input, pbar):
...
+    default_api_url = "http://localhost:8000/v3/embeddings"
+elif args["backend"] == "ovms_rerank":
+    backend_function = async_request_rerank
+    default_api_url = "http://localhost:8000/v3/rerank"
+elif args["backend"] == "tei-embed":
+    backend_function = async_request_embeddings_tei
+    default_api_url = "http://localhost:8080/embed"
+elif args["backend"] == "infinity-embeddings":
+    backend_function = async_request_embeddings
+    default_api_url = "http://localhost:7997/embeddings"
```
Comment (pgladkows) [inline demos/benchmark/embeddings/benchmark_embeddings.py:274]: "else" can be added with exception raised, when backed is different than expected

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 8]
```diff
@@ -122,6 +129,94 @@ async def async_request_embeddings(
...
+        st = time.perf_counter()
+        try:
+            async with session.post(url=api_url, json=payload,
+                                    headers=headers) as response:
+                if response.status == 200:
+                    async for chunk_bytes in response.content:
+                        if not chunk_bytes:
+                            continue
+                        #chunk_bytes = chunk_bytes.decode("utf-8")
+                        # data = json.loads(chunk_bytes)
```
[the remark this answers is not in the document]
Comment (pgladkows) [reply demos/benchmark/embeddings/benchmark_embeddings.py:159]:
```suggestion
```

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 9]
```diff
@@ -122,6 +129,94 @@ async def async_request_embeddings(
...
+                        timestamp = time.perf_counter()
+                        output.success = True
+                        output.latency =  timestamp - st
+                else:
+                    output.error = response.reason or ""
+                    output.success = False
+                    print("ERROR",response.reason)
```
Comment (pgladkows) [inline demos/benchmark/embeddings/benchmark_embeddings.py:209]:
```suggestion
                    print("ERROR", response.reason)
```

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 10]
```diff
@@ -174,4 +288,4 @@ async def limited_request_func(request_func_input, pbar):
 print("Throughput - Tokens per second:",num_tokens / benchmark_results["duration"])
 print("Mean latency: {} ms".format(round(np.mean(benchmark_results["latencies"])*1000),3))
 print("Median latency: {} ms".format(round(np.median(benchmark_results["latencies"])*1000),3))
-print("Average document length: {} tokens".format(num_tokens / len(docs)))
\ No newline at end of file
+print("Average document length: {} tokens".format(num_tokens / len(docs)))
```
Comment (pgladkows) [inline demos/benchmark/embeddings/benchmark_embeddings.py:291]: consider replacing .format() with f"" string

File: demos/rerank/README.md [hunk 11]
```diff
@@ -95,6 +118,47 @@ HF reranking: [0.99640983 0.08154089]
...
 
+## Performance benchmarking
+
+An asynchronous benchmarking client can be used to access the model server performance with various load conditions. Below are execution examples captured on dual Intel(R) Xeon(R) CPU Max 9480.
+```bash
+popd
+pushd .
+cd demos/benchmark/embeddings/
+pip install -r requirements.txt
+python benchmark_embeddings.py --backend ovms_rerank --dataset synthetic --synthetic_length 500 --request_rate inf --batch_size 20 --model BAAI/bge-reranker-large
```
Comment (ngrozae) [inline demos/rerank/README.md:129]:
```suggestion
python benchmark_embeddings.py --api_url http://localhost:8000/v3/embeddings --backend ovms_rerank --dataset synthetic --synthetic_length 500 --request_rate inf --batch_size 20 --model BAAI/bge-reranker-large 
```

File: demos/rerank/README.md [hunk 12]
```diff
@@ -95,6 +118,47 @@ HF reranking: [0.99640983 0.08154089]
...
+Number of documents: 1000
+100%|██████████████████████████████████████| 50/50 [00:19<00:00,  2.53it/s]
+Tokens: 501000
+Success rate: 100.0%. (50/50)
+Throughput - Tokens per second: 25325.17484336458
+Mean latency: 10268 ms
+Median latency: 10249 ms
+Average document length: 501.0 tokens
+
+python benchmark_embeddings.py --backend ovms_rerank --dataset synthetic --synthetic_length 500 --request_rate inf --batch_size 20 --model BAAI/bge-reranker-large
```
Comment (ngrozae) [inline demos/rerank/README.md:139]:
```suggestion
python benchmark_embeddings.py --api_url http://localhost:8000/v3/embeddings --backend ovms_rerank --dataset synthetic --synthetic_length 500 --request_rate inf --batch_size 20 --model BAAI/bge-reranker-large 
```

File: demos/rerank/README.md [hunk 13]
```diff
@@ -95,6 +118,47 @@ HF reranking: [0.99640983 0.08154089]
...
+Number of documents: 1000
+100%|██████████████████████████████████████| 50/50 [00:19<00:00,  2.53it/s]
+Tokens: 501000
+Success rate: 100.0%. (50/50)
+Throughput - Tokens per second: 25325.17484336458
+Mean latency: 10268 ms
+Median latency: 10249 ms
+Average document length: 501.0 tokens
+
+python benchmark_embeddings.py --backend ovms_rerank --dataset Cohere/wikipedia-22-12-simple-embeddings --request_rate inf
```
Comment (ngrozae) [inline demos/rerank/README.md:149]:
```suggestion
python benchmark_embeddings.py --api_url http://localhost:8000/v3/embeddings --backend ovms_rerank --dataset Cohere/wikipedia-22-12-simple-embeddings --request_rate inf 
```

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 14]
```diff
@@ -111,7 +124,95 @@ async def async_request_embeddings(
...
+                        output.latency =  timestamp - st
+                else:
+                    output.error = response.reason or ""
+                    output.success = False
+                    print("ERROR", response.reason)
+
+        except Exception:
+            output.success = False
+            exc_info = sys.exc_info()
+            output.error = "".join(traceback.format_exception(*exc_info))
```
Comment (ngrozae) [inline demos/benchmark/embeddings/benchmark_embeddings.py:177]: this part is used 3 times. maybe enclose it in a function?

Comment (dtrawins) [reply demos/benchmark/embeddings/benchmark_embeddings.py:177]: Those implementations might change for each function if we add later some processing of the returned output. So now I propose to keep is as a bit duplicated.

File: demos/benchmark/embeddings/benchmark_embeddings.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mzegla) [reply demos/benchmark/embeddings/benchmark_embeddings.py:165]: +1

File: demos/benchmark/embeddings/benchmark_embeddings.py [hunk 15]
```diff
@@ -111,7 +124,95 @@ async def async_request_embeddings(
...
+    api_url = request_func_input.api_url
+
+    async with aiohttp.ClientSession(timeout=AIOHTTP_TIMEOUT, read_bufsize=100000) as session:
+        payload = {
+            "inputs": request_func_input.documents,
+        }
+        headers = {
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {os.environ.get('OPENAI_API_KEY')}",
+        }
```
Comment (mzegla) [inline demos/benchmark/embeddings/benchmark_embeddings.py:196]: This looks like a common part. Can we extract it outside this function and use as a "const" here and in rerank?

**Review by ngrozae — APPROVED** [review]

**Review by mzegla — APPROVED** [review]

**Pull request merged**